Create openshift wrapper extension. (#7356)
quarkusio/quarkus: docs/src/main/asciidoc/openshift.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Openshift extension

include::./attributes.adoc[]

> This guide covers generating and deploying Openshift resources based on sane default and user supplied configuration.

Reviewer: ```suggestion
This guide covers generating and deploying Openshift resources based on sane defaults and user supplied configuration.
```

== Prerequisites

To complete this guide, you need:

* roughly 5 minutes
* an IDE
* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
* Apache Maven 3.5.3+
* access to an Openshift or cluster (Minishift is a viable options)

== Creating the Maven project

First, we need a new project that contains the Openshift extension. This can be done using the following command:

[source, subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=openshift-quickstart \
    -DclassName="org.acme.rest.GreetingResource" \
    -Dpath="/greeting" \
    -Dextensions="openshift"

cd openshift-quickstart
----

=== Openshift

Quarkus offers the ability to automatically generate Openshift resources based on sane defaults and user supplied configuration.
The Openshift extension is actually a wrapper extension that brings together the link:kubernetes[kubernetes] and link:container-image-s2i[container-image-s2i]

> extensions with sensible defaults so that its easier for the user to get started with Quarkus on Openshift.

Reviewer: ```suggestion
extensions with sensible defaults so that it's easier for the user to get started with Quarkus on Openshift.
```

----
mvn clean package -Dquarkus.kubernetes.deploy=true
----

The command above will trigger a container image build and will apply the generated Openshift resources, right after.
The generated resources are using Openshift's `DeploymentConfig` that is configured to automatically trigger a redeployment when a change in the `ImageStream` is noticed.
In other words, any container image build after the inital deployment will automatically trigger redeployment, without the need to delete, update or re-apply the generated resources.

=== Customizing

All available customization options are available in the link:kubenretes#openshift[Openshift configuration options].

Some examples are provided in the sections below:

==== Labels

To add a label in the generated resources:

[source]
----
quarkus.openshift.labels.foo=bar
----

==== Annotations

To add an annotation in the generated resources:

[source]
----
quarkus.openshift.annotations.foo=bar
----

==== Environment variables

To add an annotation in the generated resources:

[source]
----
quarkus.openshift.env-vars.my-env-var.value=foobar
----

The command above will add `MY_ENV_VAR=foobar` as an environment variable.

> Please, notice that the key `my-env-var` will be converted to uppercase and dashes will be replaced by underscores resulting in `MY_ENV_VAR`.

Reviewer: ```suggestion
Please note that the key `my-env-var` will be converted to uppercase and dashes will be replaced by underscores resulting in `MY_ENV_VAR`.
```

> You may also noticed that in contrast to labels, and annotations for environment variables you don't just use a key=value approach.

Reviewer: ```suggestion
You may also have noticed that in contrast to labels and annotations for environment variables you don't just use a key=value approach.
```

> That is because for environment variables there are additonal options rather than just value.

Reviewer: ```suggestion
That is because for environment variables there are additional options rather than just defining value.
```

===== Environment variables from Secret

> To add all key value pairs of a `Secret` as envrionment variables:

Reviewer: ```suggestion
To add all key value pairs of a `Secret` as environment variables:
```

[source]
----
quarkus.openshift.env-vars.my-env-var.secret=my-secret
----

===== Environment variables from ConfigMap

> To add all key value pairs of a `ConfigMap` as envrionment variables:

Reviewer: ```suggestion
To add all key value pairs of a `ConfigMap` as environment variables:
```

[source]
----
quarkus.openshift.env-vars.my-env-var.configmap=my-secret
----

==== Exposing Routes

> To expose a route for the Quarkus application:

Reviewer: ```suggestion
To expose a `Route` for the Quarkus application:
```

[source]
----
quarkus.openshift.expose=true

> ----

Reviewer 2: should we add a:

TIP: You do not need to always put these in `application.properties` you can also pass it as a command line argument: `-Dquarkus.openshift.expose=true`

Author: Agree! Let me add that.
